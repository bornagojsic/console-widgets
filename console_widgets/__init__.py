from .console_widgets import *