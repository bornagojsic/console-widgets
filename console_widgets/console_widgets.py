from errors import *
import inspect


### IDEAS
## ------------------------------------------------------------------------------------------------
## add a modify_width method which automagically modifies the width of a wigets after using the set method
## add the possibility to add a multi-line body for the widgets and other attributes
## ^ done only for the ConsoleWidget currently
## (+?) change allignment checking from COnsoleBox to another method (kinda done)
## (easy!) add Corners() class for the Box() class so that you don't need to use lists


class Box():
	def __init__(self, type: str = "Default", horizontal_border: str = "─", vertical_border: str = "│", full_intersection: str = "┼",
			left_vertical: str = "├", right_vertical: str = "┤", upper_horizontal: str = "┬", lower_horizontal: str = "┴",
			upper_left: str = "┌", upper_right: str = "┐", lower_left: str = "└", lower_right: str = "┘"):
		super(Box, self).__init__()
		
		## borders
		self.horizontal_border = horizontal_border
		self.vertical_border = vertical_border

		## intersections
		self.full_intersection = full_intersection
		self.left_vertical = left_vertical
		self.right_vertical = right_vertical
		self.upper_horizontal = upper_horizontal
		self.lower_horizontal = lower_horizontal

		## corners
		self.upper_left = upper_left
		self.upper_right = upper_right
		self.lower_right = lower_right
		self.lower_left = lower_left
		

class ConsoleWidget():
	def __init__(self, title: str = "", subtitle: str = "",
		body: str = "", width: int = None):
		self.title = title
		self.subtitle = subtitle
		self.body = body
		
		for attribute, value in self.get_attributes():
			if not type(value) is str:
				raise ParameterTypeError(value, attribute, "string")

		self.title = self.title.upper()
		self.subtitle = self.subtitle.title()

		self.parse_attributes()


	def show(self):
		for attribute in [self.title, self.subtitle, self.body]:
			if any(attribute):
				for line in attribute:
					print(line)
				print()


	def parse_attributes(self):
		## This method parses multi-line attribtes into lists
		for attribute, value in self.get_attributes():
			if type(value) is str:
				parsed_value = value.split('\n')
				setattr(self, attribute, parsed_value)


	def set(self, attr_val: dict = {}):
		if type(attr_val) is not dict:
			raise SetError(attr_val)

		[setattr(self, attribute, value) for attribute, value in attr_val.items()]

		## ! This is a temporary solution to possibly changigng the max width when resetting an attribute

		#if type(self) is ConsoleList():
		#	self.width = max(17 + len(str(len(items))), max(map(len, map(str, [self.title, self.subtitle] + self.items))))



	def get_attributes(self):
		### This code is courtesy of Matt Luongo from StackOverflow
		### https://stackoverflow.com/questions/9058305/getting-attributes-of-a-class

		attributes = inspect.getmembers(self, lambda a:not(inspect.isroutine(a)))
		return [a for a in attributes if not(a[0].startswith('__') and a[0].endswith('__'))]


class ConsoleBox(ConsoleWidget):
	def __init__(self, title: str = "", subtitle: str = "", body: str = "", box: Box = Box(), horizontal_margin: int = 0, vertical_margin: int = 0, allignment: str = "CENTER"):
		super(ConsoleBox, self).__init__()
		self.title = title
		self.subtitle = subtitle
		self.body = body
		self.box = box
		self.horizontal_margin = horizontal_margin
		self.vertical_margin = vertical_margin
		self.allignment = allignment

		self.parse_attributes()

		self.width = max(map(len, self.title + self.subtitle + self.body)) + 2 * self.horizontal_margin
		self.horizontal_border = self.box.horizontal_border * self.width


	def return_padding(self, padding):
		## The function returns the left and right paddings
		if self.allignment[0] == "CENTER":
			return (" " * ((padding // 2) + (padding % 2)), " " * (padding // 2))
		if self.allignment[0] == "LEFT":
			return ("", " " * padding)
		if self.allignment[0] == "RIGHT":
			return (" " * padding, "")


	def show(self):
		## for i in range(self.border_width):
		print(self.box.upper_left + self.horizontal_border + self.box.upper_right)

		if any(self.title):
			for title_line in self.title:
				title_padding = self.width - len(title_line)
				
				(left_t_padding, right_t_padding) = self.return_padding(title_padding)

				print(self.box.vertical_border + left_t_padding + title_line + right_t_padding + self.box.vertical_border)
				print(self.box.left_vertical + self.horizontal_border + self.box.right_vertical)

		if any(self.subtitle):
			for sub_line in self.subtitle:
				subtitle_padding = self.width - len(sub_line)
				
				(left_s_padding, right_s_padding) = self.return_padding(subtitle_padding)

				print(self.box.vertical_border + left_s_padding + sub_line + right_s_padding + self.box.vertical_border)
				print(self.box.left_vertical + self.horizontal_border + self.box.right_vertical)

		if any(self.body):
			for body_line in self.body:
				body_padding = self.width - len(body_line)
				
				(left_b_padding, right_b_padding) = self.return_padding(body_padding)

				print(self.box.vertical_border + left_b_padding + body_line + right_b_padding + self.box.vertical_border)
		print(self.box.lower_left + self.horizontal_border + self.box.lower_right)


class ConsoleList(ConsoleWidget):
	def __init__(self, title: str = "", subtitle: str = "", items: list = []):
		super(ConsoleList, self).__init__()
		self.title = title
		self.subtitle = subtitle
		self.items = items
		
		for attribute, value in self.get_attributes():
			if attribute != 'items' and not type(value) is str:
				raise ParameterTypeError(value, attribute, "string")

		self.width = max(map(len, map(str, self.title + self.subtitle + self.items)))


	def show(self):
		print("─" * self.width)
		for attribute in [self.title, self.subtitle]:
			if attribute:
				for lin in attribute:
					print(line)
				print("─" * self.width)

		for item_number, item in enumerate(self.items):
			print(f"{item_number + 1}. {item}")
		print("─" * self.width)


class ConsoleSelection(ConsoleList):
	def __init__(self, title: str = "", subtitle: str = "", items: list = []):
		self.title = title
		self.subtitle = subtitle
		self.items = items
		
		for attribute, value in self.get_attributes():
			if attribute != 'items' and not type(value) is str:
				raise ParameterTypeError(value, attribute, "string")

		self.width = max(17 + len(str(len(items))), max(map(len, map(str, [self.title, self.subtitle] + self.items))))

	def select(self):
		super(ConsoleSelection, self).show()

		selection = input(f"Select [1-{len(self.items)}]: ")
		#print("─" * self.width)

		return selection