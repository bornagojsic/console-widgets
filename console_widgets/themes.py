box_themes = {
	'default': {
		'upper_border': "─",
		'lower_border': "─",
		'left_border': "│",
		'right_border': "│",
		'horizontal_line': "─",
		'vertical_line': "│",
		'full_intersection': "┼",
		'left_vertical': "├",
		'right_vertical': "┤",
		'upper_horizontal': "┬",
		'lower_horizontal': "┴",
		'upper_left': "┌",
		'upper_right': "┐",
		'lower_left': "└",
		'lower_right': "┘",
	},

	'double': {
		'upper_border': "═",
		'lower_border': "═",
		'left_border': "║",
		'right_border': "║",
		'horizontal_line': "═",
		'vertical_line': "║",
		'full_intersection': "╬",
		'left_vertical': "╠",
		'right_vertical': "╣",
		'upper_horizontal': "╦",
		'lower_horizontal': "╩",
		'upper_left': "╔",
		'upper_right': "╗",
		'lower_left': "╚",
		'lower_right': "╝",
	},
	
	'wide': {
		'upper_border': "▄",
		'lower_border': "▀",
		'left_border': "█",
		'right_border': "█",
		'horizontal_line': "▀",
		'vertical_line': "█",
		'full_intersection': "█",
		'left_vertical': "█",
		'right_vertical': "█",
		'upper_horizontal': "█",
		'lower_horizontal': "█",
		'upper_left': "▄",
		'upper_right': "▄",
		'lower_left': "▀",
		'lower_right': "▀",
	},
	
	'dots': {
		'upper_border': ".",
		'lower_border': ".",
		'left_border': ".",
		'right_border': ".",
		'horizontal_line': ".",
		'vertical_line': ".",
		'full_intersection': ".",
		'left_vertical': ".",
		'right_vertical': ".",
		'upper_horizontal': ".",
		'lower_horizontal': ".",
		'upper_left': ".",
		'upper_right': ".",
		'lower_left': ".",
		'lower_right': ".",
	},

	'pluses': {
		'upper_border': "+",
		'lower_border': "+",
		'left_border': "+",
		'right_border': "+",
		'horizontal_line': "+",
		'vertical_line': "+",
		'full_intersection': "+",
		'left_vertical': "+",
		'right_vertical': "+",
		'upper_horizontal': "+",
		'lower_horizontal': "+",
		'upper_left': "+",
		'upper_right': "+",
		'lower_left': "+",
		'lower_right': "+",
	},

	'asterisks': {
		'upper_border': "*",
		'lower_border': "*",
		'left_border': "*",
		'right_border': "*",
		'horizontal_line': "*",
		'vertical_line': "*",
		'full_intersection': "*",
		'left_vertical': "*",
		'right_vertical': "*",
		'upper_horizontal': "*",
		'lower_horizontal': "*",
		'upper_left': "*",
		'upper_right': "*",
		'lower_left': "*",
		'lower_right': "*",
	},
}