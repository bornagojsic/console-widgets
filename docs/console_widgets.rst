console\_widgets package
========================

Submodules
----------

console\_widgets.console\_widgets module
----------------------------------------

.. automodule:: console_widgets.console_widgets
   :members:
   :undoc-members:
   :show-inheritance:

console\_widgets.errors module
------------------------------

.. automodule:: console_widgets.errors
   :members:
   :undoc-members:
   :show-inheritance:

Module contents
---------------

.. automodule:: console_widgets
   :members:
   :undoc-members:
   :show-inheritance:
