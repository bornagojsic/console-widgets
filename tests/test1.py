import sys
import os
sys.path.append(os.path.normpath(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'console_widgets')))
from console_widgets import *
import traceback


TEST_NUMBER = 1


def test(t):
	global TEST_NUMBER
	
	print("---------------------------------")

	try:
		print(f"TEST {TEST_NUMBER}")
		t()
	except Exception as exc:
		print(traceback.format_exc())

	print("---------------------------------")

	TEST_NUMBER += 1


def t1():
	c = ConsoleWidget(title="NASLOV", body="tijelo.....")

	c.show()


def t2():
	c = ConsoleWidget()

	c.show()


def t3():
	c = ConsoleWidget(title="Naslov", subtitle="cijena rada", body=25)

	c.show()


def t4():
	c = ConsoleWidget(title=[1,2], subtitle=12635, body=25.1)

	c.show()


def t5():
	c = ConsoleWidget(title="naslov", subtitle="podnaslov", body="tijelo")

	c.show()


def t6():
	c = ConsoleList(title="Lista", subtitle="borna gojsic", items=[1, 2, 3, 4, 5, 6])

	c.show()


def t7():
	c = ConsoleBox(title="naslov", subtitle="naslov2", body="tijelo....")

	c.show()


def t8():
	c = ConsoleSelection(title="Select", subtitle="Items", items=["Go right", "Go left", "Do nothing", "Go back"])

	c.select()


# [test(t) for t in [t1, t2, t3, t4, t5, t6, t7, t8]]
[test(t) for t in [t1, t2, t3, t4, t5, t6]]