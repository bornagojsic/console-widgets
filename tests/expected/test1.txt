------------------------
TEST 1
NASLOV

tijelo.....

------------------------
------------------------
TEST 2
------------------------
------------------------
TEST 3
Traceback (most recent call last):
  File "F:\Programming\github\console-widgets\tests\test1.py", line 18, in test
    t()
  File "F:\Programming\github\console-widgets\tests\test1.py", line 40, in t3
    c = ConsoleWidget(title="Naslov", subtitle="cijena rada", body=25)
  File "F:\Programming\github\console-widgets\console_widgets\console_widgets.py", line 31, in __init__
    raise ParameterTypeError(value, attribute, "string")
errors.ParameterTypeError: 25 is of type int. -> The body can only be of type string!

------------------------
------------------------
TEST 4
Traceback (most recent call last):
  File "F:\Programming\github\console-widgets\tests\test1.py", line 18, in test
    t()
  File "F:\Programming\github\console-widgets\tests\test1.py", line 46, in t4
    c = ConsoleWidget(title=[1,2], subtitle=12635, body=25.1)
  File "F:\Programming\github\console-widgets\console_widgets\console_widgets.py", line 31, in __init__
    raise ParameterTypeError(value, attribute, "string")
errors.ParameterTypeError: 25.1 is of type float. -> The body can only be of type string!

------------------------
------------------------
TEST 5
NASLOV

Podnaslov

tijelo

------------------------
------------------------
TEST 6
Traceback (most recent call last):
  File "F:\Programming\github\console-widgets\tests\test1.py", line 18, in test
    t()
  File "F:\Programming\github\console-widgets\tests\test1.py", line 58, in t6
    c = ConsoleList(title="Lista", subtitle="borna gojsic", items=[1,2,3,4,5,6])
  File "F:\Programming\github\console-widgets\console_widgets\console_widgets.py", line 144, in __init__
    raise ParameterTypeError(value, attribute, "string")
errors.ParameterTypeError: [''] is of type list. -> The body can only be of type string!

------------------------
